### PR TITLE
[LoongArch] Optimize codegen for ISD::{ROTL,ROTR}

### PR DESCRIPTION
**Pull request by heiher**

### CODE REVIEW

**Review by SixWeining — COMMENTED** [review]
For the title, should we mention ISD::ROTR as well?

File: llvm/test/CodeGen/LoongArch/rotl-rotr.ll
[the remark this answers is not in the document]
Comment (heiher) [reply llvm/test/CodeGen/LoongArch/rotl-rotr.ll:1]: Removed. Thanks.

File: llvm/lib/Target/LoongArch/LoongArchISelLowering.cpp [hunk 1]
```diff
@@ -79,6 +79,7 @@ LoongArchTargetLowering::LoongArchTargetLowering(const TargetMachine &TM,
   setOperationAction(ISD::SRL_PARTS, GRLenVT, Custom);
   setOperationAction(ISD::FP_TO_SINT, GRLenVT, Custom);
   setOperationAction(ISD::ROTL, GRLenVT, Expand);
+  setOperationAction(ISD::ROTR, GRLenVT, Legal);
```
Comment (wangleiat) [inline llvm/lib/Target/LoongArch/LoongArchISelLowering.cpp:82]: This operation might be redundant, as `ISD::ROTR` defaults to being legal.

File: llvm/lib/Target/LoongArch/LoongArchISelLowering.cpp [hunk 2]
```diff
@@ -1715,6 +1712,26 @@ static SDValue customLegalizeToWOp(SDNode *N, SelectionDAG &DAG, int NumOp,
   return DAG.getNode(ISD::TRUNCATE, DL, N->getValueType(0), NewRes);
 }
 
+static SDValue customLegalizeToRotateWOp(SDNode *N, SelectionDAG &DAG,
```
Comment (wangleiat) [inline llvm/lib/Target/LoongArch/LoongArchISelLowering.cpp:1715]: Can we try to reuse  `customLegalizeToWOp`?

File: llvm/lib/Target/LoongArch/LoongArchInstrInfo.td [hunk 3]
```diff
@@ -1116,12 +1115,10 @@ def : PatGprGpr<srem, MOD_D>;
...
 def : PatGprGpr<rotr, ROTR_D>;
 def : PatGprGpr<loongarch_rotr_w, ROTR_W>;
+def : PatGprGpr_32<rotr, ROTR_W>;
 def : PatGprImm<rotr, ROTRI_D, uimm6>;
 def : PatGprImm_32<rotr, ROTRI_W, uimm5>;
-def : Pat<(loongarch_rotl_w GPR:$rj, uimm5:$imm),
-          (ROTRI_W GPR:$rj, (ImmSubFrom32 uimm5:$imm))>;
-def : Pat<(sext_inreg (loongarch_rotl_w GPR:$rj, uimm5:$imm), i32),
-          (ROTRI_W GPR:$rj, (ImmSubFrom32 uimm5:$imm))>;
+def : PatGprImm_32<loongarch_rotr_w, ROTRI_W, uimm5>;
```
Comment (wangleiat) [inline llvm/lib/Target/LoongArch/LoongArchInstrInfo.td:1121]: def : PatGprImm<loongarch_rotr_w, ROTRI_W, uimm5>;

Comment (heiher) [reply llvm/lib/Target/LoongArch/LoongArchInstrInfo.td]: All done. Thanks.

**Review by gonglingqin — APPROVED** [review]

**Review by wangleiat — APPROVED** [review]

**Pull request merged**